AFQ/api.py

```diff
@@ -122,59 +154,165 @@ def __init__(self, raw_path=None, preproc_path=None,
+            self.data_frame['dwi_file'].apply(_get_affine)
+
+    def get_dwi_affine(self):
+        return self.data_frame['dwi_affine']
+
+    dwi_affine = property(get_dwi_affine, set_dwi_affine)
+
+    def __getitem__(self, k):
+        return self.data_frame.__getitem__(k)
+
+    def _brain_extract(self, row, median_radius=4, numpass=4, autocrop=False,
+                       vol_idx=None, dilate=None, force_recompute=False):
+        if not op.exists(row['brain_mask_file']) or force_recompute:
+            self.set_dwi_data_img()
+            img = row['dwi_data_img']
+            data = img.get_data()
+            gtab = row['gtab']
+            mean_b0 = np.mean(data[..., ~gtab.b0s_mask], -1)
+            _, brain_mask = median_otsu(mean_b0, median_radius, numpass,
+                                        autocrop, dilate=dilate)
+            be_img = nib.Nifti1Image(brain_mask.astype(int),
+                                     img.affine)
+            nib.save(be_img, row['brain_mask_file'])
+            return be_img
+        else:
+            return nib.load(row['brain_mask_file'])
+
+    def set_brain_mask(self, median_radius=4, numpass=4, autocrop=False,
+                       vol_idx=None, dilate=None, force_recompute=False):
+        if 'brain_mask_img' not in self.data_frame.columns or force_recompute:
             self.data_frame['brain_mask_file'] =\
-                self.data_frame.apply(_brain_extract_fname, axis=1)
+                self.data_frame.apply(_get_fname, suffix='_brain_mask.nii.gz',
+                                      axis=1)
 
             self.data_frame['brain_mask_img'] =\
-                self.data_frame.apply(_brain_extract, axis=1,
+                self.data_frame.apply(self._brain_extract,
+                                      axis=1,
                                       median_radius=median_radius,
                                       numpass=numpass,
                                       autocrop=autocrop,
                                       vol_idx=vol_idx,
                                       dilate=dilate,
                                       force_recompute=force_recompute)
 
-    def fit_tensors():
-        raise NotImplementedError
+    def get_brain_mask(self):
+        self.set_brain_mask()
+        self.data_frame['brain_mask'] =\
+            self.data_frame['brain_mask_img'].apply(nib.Nifti1Image.get_data)
+        return self.data_frame['brain_mask']
+
+    brain_mask = property(get_brain_mask, set_brain_mask)
+
+    def set_dwi_data_img(self):
+        if 'dwi_data_img' not in self.data_frame.columns:
+            self.data_frame['dwi_data_img'] =\
+                self.data_frame['dwi_file'].apply(nib.load)
+
+    def get_dwi_data_img(self):
+        self.set_dwi_data_img()
+        return self.data_frame['dwi_data_img']
+
+    dwi_data_img = property(get_dwi_data_img, set_dwi_data_img)
+
+    def _dti(self, row, mask=None):
+        if not op.exists('dti_params_file'):
```

Comment:
I'm not sure we want this check. Since this is done at a level higher as well, we should assume at this point that you want to recompute and overwrite. Note that the higher level function had an overwrite option and this check will undo that overwrite option